drivers/bluetooth/hci/rpmsg.c

```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <init.h>
+#include <sys/byteorder.h>
+
+#include <bluetooth/bluetooth.h>
+#include <bluetooth/hci.h>
+#include <bluetooth/hci_driver.h>
+
+#define BT_DBG_ENABLED IS_ENABLED(CONFIG_BT_DEBUG_HCI_DRIVER)
+#define LOG_MODULE_NAME bt_hci_driver
+#include "common/log.h"
+
+#define RPMSG_CMD 0x01
+#define RPMSG_ACL 0x02
+#define RPMSG_SCO 0x03
+#define RPMSG_EVT 0x04
+
+int bt_rpmsg_platform_init(void);
+int bt_rpmsg_platform_send(struct net_buf *buf);
+
+static inline void bt_buf_set_prio(struct net_buf *buf, bool prio)
+{
+	*((u8_t *)net_buf_user_data(buf) + 1) = prio;
+}
+
+static inline bool bt_buf_get_prio(struct net_buf *buf)
+{
+	return (bool)(*((u8_t *)net_buf_user_data(buf) + 1));
+}
+
+static struct net_buf *bt_rpmsg_evt_recv(u8_t *data, size_t remaining)
+{
+	struct bt_hci_evt_hdr hdr;
+	struct net_buf *buf;
+
+	if (remaining < sizeof(hdr)) {
+		BT_ERR("Not enough data for event header");
+		return NULL;
+	}
+
+	memcpy((void *)&hdr, data, sizeof(hdr));
+	data += sizeof(hdr);
+	remaining -= sizeof(hdr);
+
+	if (remaining != hdr.len) {
+		BT_ERR("Event payload length is not correct");
+		return NULL;
+	}
+	BT_DBG("len %u", hdr.len);
+
+	buf = bt_buf_get_evt(hdr.evt, false, K_NO_WAIT);
+	if (!buf) {
+		BT_ERR("No available event buffers!");
+		return buf;
+	}
+
+	net_buf_add_mem(buf, &hdr, sizeof(hdr));
+	bt_buf_set_prio(buf, bt_hci_evt_is_prio(hdr.evt));
+
+	net_buf_add_mem(buf, data, remaining);
+
+	return buf;
+}
+
+static struct net_buf *bt_rpmsg_acl_recv(u8_t *data, size_t remaining)
+{
+	struct bt_hci_acl_hdr hdr;
+	struct net_buf *buf;
+
+	if (remaining < sizeof(hdr)) {
+		BT_ERR("Not enough data for ACL header");
+		return NULL;
+	}
+
+	buf = bt_buf_get_rx(BT_BUF_ACL_IN, K_NO_WAIT);
+	if (buf) {
+		memcpy((void *)&hdr, data, sizeof(hdr));
+		data += sizeof(hdr);
+		remaining -= sizeof(hdr);
+
+		net_buf_add_mem(buf, &hdr, sizeof(hdr));
+		bt_buf_set_prio(buf, false);
+	} else {
+		BT_ERR("No available ACL buffers!");
+	}
+
+	if (remaining != sys_le16_to_cpu(hdr.len)) {
+		BT_ERR("ACL payload length is not correct");
+		return NULL;
```

Comment:
There's a buffer leak here. You should be doing `net_buf_unref(buf)` before returning

Reply:
fixed